sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/package.scala

```diff
@@ -45,18 +51,101 @@ package object analysis {
       throw new AnalysisException(msg, t.origin.line, t.origin.startPosition)
     }
 
-    /** Fails the analysis at the point where a specific tree node was parsed. */
+    /** Fails the analysis at the point where a specific tree node was parsed with a given cause. */
     def failAnalysis(msg: String, cause: Throwable): Nothing = {
       throw new AnalysisException(msg, t.origin.line, t.origin.startPosition, cause = Some(cause))
     }
 
+    /**
+     * Fails the analysis at the point where a specific tree node was parsed using a provided
+     * error class and message parameters.
+     */
     def failAnalysis(errorClass: String, messageParameters: Map[String, String]): Nothing = {
       throw new AnalysisException(
         errorClass = errorClass,
         messageParameters = messageParameters,
         origin = t.origin)
     }
 
+    /**
+     * Fails the analysis at the point where a specific tree node was parsed using a provided
+     * error class and subclass and message parameters.
+     */
+    def failAnalysis(
+        errorClass: String,
+        errorSubClass: String,
+        messageParameters: Map[String, String] = Map.empty[String, String]): Nothing = {
+      throw new AnalysisException(
+        errorClass = errorClass,
+        errorSubClass = errorSubClass,
+        messageParameters = messageParameters,
+        origin = t.origin)
+    }
+
+    /**
+     * Fails the analysis at the point where a specific tree node was parsed using a provided
+     * error class and subclass and one message parameter comprising a plan string. The plan string
+     * will be printed in the error message if and only if the corresponding Spark configuration is
+     * enabled.
+     */
+    def failAnalysis(
+        errorClass: String,
+        errorSubClass: String,
+        treeNodes: Seq[TreeNode[_]]): Nothing = {
+      // Normalize expression IDs in the query plan to keep tests deterministic.
```

Comment:
Shall we use `Expression.canonicalized` instead? There are existing methods for such purposes.

Reply:
I tried this, but this `canonicalized` method is not available during analysis since it uses `transformUp`. Therefore a separate implementation is needed.

Reply:
![image](https://user-images.githubusercontent.com/99207096/190524824-372496a8-7a6f-464d-9239-57df10ecd168.png)

Reply:
Well, at least I would suggest moving these methods out of this file. We can do it in `Expression` or other Util objects.

Reply:
Hi @gengliangwang I looked at this at length some more, and I was able to reuse the existing plan `canonicalize` method after all using `AnalysisHelper.allowInvokingTransformsInAnalyzer` in one specific place instead. Now there is no need to introduce new code for this purpose. Please take another look.